src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs

```diff
@@ -117,15 +122,15 @@ public DirectoryLoadContext(string directory, DefaultAnalyzerAssemblyLoader load
                 {
                     // Delegate to the compiler's load context to load the compiler or anything
                     // referenced by the compiler
-                    return null;
+                    return _compilerLoadContext.LoadFromAssemblyName(assemblyName);
                 }
 
                 var assemblyPath = Path.Combine(Directory, simpleName + ".dll");
                 if (!_loader.IsKnownDependencyLocation(assemblyPath))
                 {
                     // The analyzer didn't explicitly register this dependency. Most likely the
                     // assembly we're trying to load here is netstandard or a similar framework
-                    // assembly. We assume that if that is not the case, then the parent ALC will
+                    // assembly. We assume that if that is not the case, then the default ALC will
                     // fail to load this.
                     return null;
```

Comment:
```suggestion
                    _compilerLoadContext.LoadFromAssemblyName(assemblyName);
```

This also needs to fall back to the compiler ALC

Reply:
Could you please elaborate on what incorrect behavior could arise from the absence of this change? I think we need a test which fails in the absence of this suggested change. Do you have any thoughts on what such a test looks like?

Reply:
Let's identify the different ALC that are in play here: 

1. Compiler ALC: this is the ALC which contains the compiler binaries
2. Default ALC: this is the `AssemblyLoadContext.Default` value 
3. Analyzer ALC: this is the set of ALC loaded for analyzers

In the case of the compiler 1 & 2 are the same. But in the case of VS 1 & 2 are separate with 2 being the service hub process and 1 being the ALC it creates for loading services. 

Lets now consider a case where an analyzer is built against `netstandard1.3`. I chose that target framework because it's different than `netstandard2.0` which is what the compiler typically builds against. That means they will tend to see a different set of references at compile time. Hence we can get into situations where the analyzer has references in its list which don't appear in our `CompilerAssemblySimpleNames` collection. A concrete example of this is `System.Threading.Tasks.Extensions.dll`. That is not included in `CompilerAssemblySimpleNames` but it is a reference that is used by analyzers templates. 

Consider the case of an analyzer that references `System.Threading.Tasks.Extensions.dll` and uses `ValueTask<T>` in one of its methods. When that method is used it will begin with a call to `LoadFromAssemblyName` for `System.Threading.Tasks.Extensions`. That is not in `CompilerAssemblySimpleNames` nor will it be in a known dependency location. Hence it will hit this line and return `null`

The return of `null` means that the load goes back to the default ALC (2 above). Hence between the compiler and VS we end up with two different results. In the compiler it is loaded into the default + compiler ALC as its the same item. In VS the assembly gets loaded into the service hub but does not get loaded into the compiler ALC. 

For types that are never communicated across the compiler API that likely isn't much of a problem. It's an implementation detail of the analyzer so not important if the type isn't accessible in the compiler. I'm guessing if we dig through the facades enough though we will find a case where this results in a type in the compiler public API not being loaded into the compiler ALC. The only guard we have against that today is `CompilerAssemblySimpleNames` and that is an imperfect list. It represents the set of assemblies the compiler knew about at **its** compile time. When an analyzer builds it could have a very different view of the world because types move between assemblies from release to release. It's a case where we can only prove we're safe by exhaustively looking through all TFM and ensuring that types didn't move in such a way that we'd get broken here. I'd prefer to just assume that it probably can happen and guard against it. 

Overall I can't think of any case where we should be loading a managed assembly into the default ALC. It should either be loaded into the analyzer ALC or the compiler ALC. Any time where we fall back to the default ALC we're essentially depending on our host to do the correct thing which isn't reasonable.

Reply:
After writing all that part of me wants to suggest we try out deleting `CompilerAssemblySimpleNames`, default to loading into the compiler ALC and then and only then fall back to analyzer ALC if the dependency is in the known dependency list. It's a much bigger change but I'm starting to think it's the correct one. 

Should probably move forward with the direction of this PR though, merge so we unblock VS first though. But still think we need to change this line before merging.